File: core/src/main/java/io/github/msameer0/rhythmicrush/game/gameplay/players/Cube.java
package io.github.msameer0.rhythmicrush.game.gameplay.players;

import io.github.msameer0.rhythmicrush.game.registries.Registry;

/**
 * Represents a specific player character implementation that follows a "Cube" mechanic.
 * The Cube is affected by gravity and can perform a single jump when grounded.
 * This class supports object pooling through its no-arg constructor and {@link #init(float, float)} method.
 */
@Registry(id = "cube")
public class Cube extends AbstractPlayer {
    public float gravity = -1800f;
    public float jumpVelocity = 600f;
    private boolean jumpHeld = false;
    private boolean isGrounded = false;

    public Cube(float startX, float groundY) {
        super(startX, groundY);
        this.type = PlayerType.CUBE;
    }

    /**
     * No-arg constructor for pooling — call init() before use.
     */
    public Cube() {
        super(0, 0);
        this.type = PlayerType.CUBE;
    }

    /**
     * Reinitialise this Cube for reuse from the pool.
     */
    @Override
    public Cube init(float startX, float startY, float velocityY, boolean jumpHeld) {
        this.type = PlayerType.CUBE;
        x = startX;
        y = startY;
        this.velocityY = velocityY;
        isGrounded = false;
        this.jumpHeld = jumpHeld;
        this.gravityFlipped = false;
        world = null;
        bounds.setPosition(x, y);
        return this;
    }

    /**
     * Legacy init for backward compatibility.
     */
    @Override
    public Cube init(float startX, float startY) {
        return init(startX, startY, 0, false);
    }

    @Override
    public void update(float delta, float groundY) {
        isGrounded = false;
        float effectiveGravity = gravityFlipped ? -gravity : gravity;
        velocityY += effectiveGravity * delta;
        y += velocityY * delta;

        if (!gravityFlipped) {
            if (y <= groundY) {
                y = groundY;
                velocityY = 0;
                isGrounded = true;
            }
        } else {
            // If we want a ceiling limit, we could add it here.
            // For now, standard ground behavior is only for normal gravity.
        }
        updateBounds();
    }

    @Override
    public void jump() {
        if (isGrounded) {
            velocityY = gravityFlipped ? -jumpVelocity : jumpVelocity;
            isGrounded = false;
        }
    }

    @Override
    public void setJumpHeld(boolean held) {
        this.jumpHeld = held;
    }

    @Override
    public boolean isJumpHeld() {
        return jumpHeld;
    }

    @Override
    public void setGrounded(boolean g) {
        this.isGrounded = g;
    }

    @Override
    public void tryJump() {
        if (jumpHeld && isGrounded) jump();
    }

    @Override
    public boolean isGrounded() {
        return isGrounded;
    }

    @Override
    public void copyState(AbstractPlayer other) {
            this.x = other.x;
            this.y = other.y;
            this.velocityY = other.velocityY;
            this.isGrounded = other.isGrounded();
            this.jumpHeld = other.isJumpHeld();
            this.gravityFlipped = other.isGravityFlipped();
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/game/gameplay/players/Cube.kt
package io.github.msameer0.rhythmicrush.game.gameplay.players

import io.github.msameer0.rhythmicrush.game.registries.Registry

/**
 * Standard Cube game mode focusing on platforming and precise jumping.
 */
@Registry(id = "cube")
class Cube : AbstractPlayer {

    @JvmField
    var jumpVelocity: Float = 1200f

    private var isGrounded: Boolean = false

    private var coyoteTimer: Float = 0f

    companion object {
        private const val COYOTE_TIME = 0.083f
    }

    constructor(startX: Float, groundY: Float) : super() {
        this.x = startX
        this.y = groundY
        gravity = -3600f
        type = PlayerType.CUBE
    }

    constructor() : super() {
        gravity = -3600f
        type = PlayerType.CUBE
    }

    override fun init(startX: Float, startY: Float, velocityY: Float, jumpHeld: Boolean): Cube {
        type = PlayerType.CUBE
        x = startX
        y = startY
        gravity = -3600f
        this.velocityY = velocityY
        isGrounded = false
        this.jumpHeld = jumpHeld
        jumpConsumed = false
        gravityFlipped = false
        setMini(false)
        coyoteTimer = 0f
        world = null
        bounds.setPosition(x, y)
        return this
    }

    override fun init(startX: Float, startY: Float): Cube = init(startX, startY, 0f, false)

    override fun update(delta: Float, groundY: Float) {
        val wasGrounded = isGrounded
        isGrounded = false
        currentSlopeRotation = 0f

        val effectiveGravity = if (gravityFlipped) -gravity else gravity
        velocityY += effectiveGravity * delta
        y += velocityY * delta

        if (!gravityFlipped) {
            if (!isGrounded && y <= groundY) {
                y = groundY
                velocityY = 0f
                isGrounded = true
            }
        } else {
            @Suppress("UNUSED_VARIABLE")
            val ceilingY = 1080 - groundY - height
        }

        if (isGrounded) {
            coyoteTimer = COYOTE_TIME
        } else if (!wasGrounded) {
            coyoteTimer = maxOf(0f, coyoteTimer - delta)
        }

        updateBounds()
    }

    private fun canJump(): Boolean = isGrounded || coyoteTimer > 0f

    override fun jump() {
        if (canJump()) {
            val v = if (mini) jumpVelocity * 0.75f else jumpVelocity
            velocityY = if (gravityFlipped) -v else v
            isGrounded = false
            coyoteTimer = 0f
            jumpConsumed = true
        }
    }

    override fun tryJump() {
        if (jumpHeld && canJump()) jump()
    }

    override fun setGrounded(g: Boolean) {
        isGrounded = g
    }

    override fun isGrounded(): Boolean = isGrounded

    override fun copyState(other: AbstractPlayer) {
        x = other.x
        y = other.y
        velocityY = other.velocityY
        isGrounded = other.isGrounded()
        jumpHeld = other.isJumpHeld()
        jumpConsumed = other.isJumpConsumed()
        gravityFlipped = other.isGravityFlipped()
        currentSlopeRotation = other.getCurrentSlopeRotation()
        setMini(other.isMini())
        coyoteTimer = 0f
    }
}
